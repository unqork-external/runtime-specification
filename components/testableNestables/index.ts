export * from './testableNestables.definition'
export * from './testableNestables.nestables'
