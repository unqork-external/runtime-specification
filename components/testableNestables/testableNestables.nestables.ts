import { Description, Property } from '@tsed/schema'

import { Nestable, NestableDef, NestableType } from '../../nestables'

export enum NestablesId {
  ARRAY = 'arrayNestables',
  MATRIX = 'matrixNestables',
}

export class ArrayNestables extends Nestable {
  @Property()
  propertyName: string = NestablesId.ARRAY

  @Property()
  type: NestableType = NestableType.ARRAY

  @Property()
  children: string = 'childrenInArray'
}

export class MatrixNestables extends Nestable {
  @Property()
  propertyName: string = NestablesId.MATRIX

  @Property()
  type: NestableType = NestableType.MATRIX

  @Property()
  children: string = 'matrixRowStructure'
}

export class TestableNestablesNestables extends NestableDef {
  @Description('Array-style nestable data')
  arrayNestables = new ArrayNestables()

  @Description('Matrix-style nestable data')
  matrixNestables = new MatrixNestables()
}
