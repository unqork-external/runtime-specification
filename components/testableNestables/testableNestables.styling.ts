import { Description, Property } from '@tsed/schema'

import { TestableNestablesTargets } from './testableNestablesTargets.enum'
import { Css } from '../../../decorators/schema/css.decorator'
import { type CssClassOrProps } from '../../styling/cssClassOrProps.type'
import { StylingModel } from '../../styling/styling.model'
import { TargetsModel } from '../../viewTargets/targets.model'

export class TestableNestablesStyleTargets implements TargetsModel<TestableNestablesTargets> {
  @Css()
  @Description('Adds styles to testableNestables root ComponentWrapper.')
  [TestableNestablesTargets.root]: CssClassOrProps
}

export class TestableNestablesStyling extends StylingModel<TestableNestablesStyleTargets> {
  @Property()
  declare targets: TestableNestablesStyleTargets
}
