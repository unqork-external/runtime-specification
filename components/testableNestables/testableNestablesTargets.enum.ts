export enum TestableNestablesTargets {
  root = 'root',
}
