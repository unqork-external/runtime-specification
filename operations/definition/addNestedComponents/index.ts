export * from './addNestedComponents.model'
export * from './addNestedComponents.options'
