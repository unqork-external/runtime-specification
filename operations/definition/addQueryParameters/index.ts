export * from './addQueryParameters.model'
export * from './addQueryParameters.options'
