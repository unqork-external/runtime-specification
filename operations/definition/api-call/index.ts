export * from './apiCall.type'
export * from './method.type'
export * from './serviceType.type'
export * from './apiCall.model'
export * from './apiCall.options'
