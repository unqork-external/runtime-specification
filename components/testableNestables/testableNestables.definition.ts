import { CollectionOf, Const, Default, Description, DiscriminatorValue, Optional, Required } from '@tsed/schema'

import { TestableNestablesNestables } from './testableNestables.nestables'
import { TestableNestablesStyling } from './testableNestables.styling'
import { TestableNestablesTargets } from './testableNestablesTargets.enum'
import { TrimmedDescription, ViewTargets } from '../../../decorators'
import { BaseComponentDefinition } from '../../base-component-interface'
import { Display } from '../../component-composition'
import { NestableType } from '../../nestables'
import type { SignalTargets } from '../../signals'

@TrimmedDescription(`
  The Testable Nestables component is a barebones "nestables" component created to isolate and test "nestables" functionality.
`)
@DiscriminatorValue('testableNestables')
@ViewTargets(TestableNestablesTargets)
export class TestableNestablesComponentDefinition extends BaseComponentDefinition {
  @Const('testableNestables')
  type: 'testableNestables' = 'testableNestables' as const

  @Optional()
  display: Display = new Display()

  @Optional()
  @Default(NestableType.ARRAY)
  @Description('Current nestable type that the component should implement')
  nestableType: NestableType = NestableType.ARRAY

  @Optional()
  @Description('Array-style nestable data')
  nestables = new TestableNestablesNestables()

  @CollectionOf(BaseComponentDefinition)
  @Description('Child components for the array nestable')
  childrenInArray: BaseComponentDefinition[]

  @CollectionOf(BaseComponentDefinition)
  @Description('Single array structure that defines what a row within the matrix should look like')
  matrixRowStructure: BaseComponentDefinition[]

  @Optional()
  @Description('Two dimensional matrix data to add on component initialization')
  initialMatrixData?: Array<BaseComponentDefinition[]>

  @Optional()
  @Description('Data object to add to component as nestable children')
  addChildrenData?: BaseComponentDefinition[] | Array<BaseComponentDefinition[]>

  @Optional()
  declare styling?: TestableNestablesStyling

  @Optional()
  declare signals?: SignalTargets<TestableNestablesTargets>
}

export class TestableNestablesComponentState extends TestableNestablesComponentDefinition {
  @Required()
  @Description('Child ids for all array nestable references within the component.')
  arrayNestables: string[] = []

  @Required()
  @Description('Child ids for all matrix nestable references within the component.')
  matrixNestables: string[] = []
}
