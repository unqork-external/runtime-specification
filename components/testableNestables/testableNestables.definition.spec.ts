import { TestableNestablesComponentDefinition } from './testableNestables.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('TestableNestables Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(TestableNestablesComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
  })
})
