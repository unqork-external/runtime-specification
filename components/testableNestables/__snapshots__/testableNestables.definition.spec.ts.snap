// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestableNestables Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
]
`;

exports[`TestableNestables Contract Tests should ensure the contract is clear 2`] = `
{
  "addChildrenData": {
    "description": "Data object to add to component as nestable children",
    "type": "array",
  },
  "childrenInArray": {
    "description": "Child components for the array nestable",
    "items": {
      "$ref": "#/definitions/BaseComponentDefinition",
    },
    "type": "array",
  },
  "display": {
    "$ref": "#/definitions/Display",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "initialMatrixData": {
    "description": "Two dimensional matrix data to add on component initialization",
    "type": "array",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "matrixRowStructure": {
    "description": "Single array structure that defines what a row within the matrix should look like",
    "items": {
      "$ref": "#/definitions/BaseComponentDefinition",
    },
    "type": "array",
  },
  "nestableType": {
    "default": "array",
    "description": "Current nestable type that the component should implement",
    "type": "string",
  },
  "nestables": {
    "description": "Array-style nestable data",
    "type": "object",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/TestableNestablesStyling",
  },
  "type": {
    "const": "testableNestables",
    "description": "Name of the type of component.",
    "examples": [
      "testableNestables",
    ],
    "minLength": 1,
    "type": "string",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "Current value of the component",
    "type": [
      "null",
      "integer",
      "number",
      "string",
      "boolean",
      "array",
      "object",
    ],
  },
}
`;
